src/System.Linq/src/System/Linq/First.cs

```diff
@@ -89,12 +89,40 @@ private static TSource TryGetFirst<TSource>(this IEnumerable<TSource> source, Fu
                 return ordered.TryGetFirst(predicate, out found);
             }
 
-            foreach (TSource element in source)
+            if (source is TSource[] array)
             {
-                if (predicate(element))
+                foreach (TSource element in array)
                 {
-                    found = true;
-                    return element;
+                    if (predicate(element))
+                    {
+                        found = true;
+                        return element;
+                    }
+                }
+            }
+            else if (source is List<TSource> list)
+            {
+                TSource element;
+                for (int i = 0; i < list.Count; ++i)
+                {
+                    element = list[i];
+
+                    if (predicate(element))
+                    {
+                        found = true;
+                        return element;
+                    }
+                }
+            }
+            else
+            {
+                foreach (TSource element in source)
+                {
+                    if (predicate(element))
+                    {
+                        found = true;
+                        return element;
+                    }
```

Comment:
Can you please share real-world examples where these special cases help? I get that they improve microbenchmarks, by avoiding the enumerator allocation and interface calls in the case of an array/list.  I'm trying to understand in what circumstance that improvement is beneficial enough to warrant the additional code.  It's not a ton of additional code, but it is additional code (in generic methods), additional type checks (which do have a negative impact on non-arrays/lists), requiring additional maintenance, etc.  Can you point to real-world examples on GitHub or elsewhere where the throughput this provides is measurable and meaningful in context?  Thanks!

Reply:
In my experience a lot of business software relies heavily on LINQ, for good and bad.
A large portion of this will ofcourse be on EF or LINQ to Enties, which is good.
However, when constructs are encountered that can not be translated by the ORM things will be pulled in memory, usually by a call to `ToList()`.

Combine this with code quality analysis with SonarQube, which has [a rule (S2971)](https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs#L416) that encourages dropping `Where` and moving the predicate into whatever is chained.
These calls will be marked with as a 'Major' code smell, most likely forcing a fix. 
Which leads me to believe there is a lot of code that has the exact code construct that this PR optimizes for the most commonly used in memory collections.

The funny thing with that sonar rule is that it also encourages dropping calls to `ToArray()` and `ToList()` in LINQ queries, suggesting `AsEnumerable()` to transfer processing into memory. But this does not work with EF 6 (not sure about 6.3 though) and LINQ to Entities, in my experience.

Another area where LINQ on arrays in particular is used a lot, is in code that performs reflection.
Since almost all reflection API's (`GetInterfaces()`, `GetMethods()` etc.) return arrays.
Reflection code is slow enough as it is, and 'allocaty' enough without using LINQ. 

For corefx it might be worthwhile to have these kind of PRs, but they do demonstrate how easy these LINQ calls sneak in and require active optimization:
https://github.com/dotnet/corefx/pull/41566/files (last file)
https://github.com/dotnet/corefx/pull/41552/files (2nd file)

Some example usages in which allocations can be reduced by this PR, or are improved automatically:
https://github.com/dotnet/corefx/blob/0cc22ccba707d7552159363d586262fa9c4e8a2a/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CollectionDataContract.cs#L1012
https://github.com/dotnet/corefx/blob/0cc22ccba707d7552159363d586262fa9c4e8a2a/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CollectionDataContract.cs#L1301
https://github.com/PowerShell/PowerShell/blob/bf912460299b1920b5f9684cfd9acd18e03d0153/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#L428

Reply:
Thank you for the response.

> Reflection code is slow enough as it is, and 'allocaty' enough without using LINQ.

This is my point.  Does this change actually move the needle?  Seems like it wouldn't.

Reply:
I'm not so sure either, I'm having a hard time finding code on GitHub in big projects where these changes make a *meaningful* impact.